contracts/Account.cairo

```diff
@@ -194,6 +205,34 @@ func execute{
     return (response_len=response.retdata_size, response=response.retdata)
 end
 
+#
+# Private functions
+#
+
+func _set_public_key{
+        syscall_ptr : felt*, 
+        pedersen_ptr : HashBuiltin*,
+        ecdsa_ptr: SignatureBuiltin*,
+        range_check_ptr
+    }(
+        new_public_key: felt,
+        signature_len: felt,
+        signature: felt*
+    ):
+    # first we set it so `is_valid_signature` can pick it up
+    public_key.write(new_public_key)
+
+    let hash_ptr = pedersen_ptr
+    with hash_ptr:
+        # check key ownership of the key to prevent spoofing
+        # validate a signature on this contract address to prevent replays
+        let (self) = get_contract_address()
+        let (identity_hash) = hash2(self, 0)
+        is_valid_signature(identity_hash, signature_len, signature)
```

Comment:
I don't think that works. Say we use this authentication mechanism to initialize a rightful account, what prevents a malicious actor to replay that very message/signature on another Account?

Reply:
Hmm good point. I'm having trouble considering a good approach here. I think your initial idea is on the right track since the `identity_hash` will always be different, but the signature does not work. I'm curious if I'm missing something from your approach and/or if you have other ideas for this fix

Reply:
What about including an `account_confirmed` bool variable that only gets set to `TRUE` upon signing a transaction? We can include an `assert` with `get_public_key` that ensures the queried pubkey has proved their ownership